### PR TITLE
adding support for Multiplex SRXL V2

### PR DESCRIPTION
**Pull request by saperlot**
Reopened Pull Request #1538 
(i had to delete the branch so it has closed)

SRXL implementation to a seperate file,  …
improved code within ISR
CRC code outsourced
docs updated

please also merge the configurator pull request

### CODE REVIEW

**Comment by martinbudden** [comment]
Why not define a specific crc `crc16_CCITT`? This is the CRC that is used in the other places in Cleanflight (apart from simple checksums, which couldn't use the generic CRC you have defined anyway). That is:
```
uint16_t crc16_CCITT(uint16_t crc, uint16_t value)
{
    uint8_t i;

    crc = crc ^ value << 8;

    for (int ii = 0; ii < 8; ++ii) {
        if (crc & 0x8000) {
            crc = crc << 1 ^ 0x1021;
        } else {
            crc = crc << 1;
        }
    }
    return crc;
}
```

**Comment by saperlot** [comment]
added it as CRC CCITT but still with uint8_t as value, SUMD also use 8 bit as input, crc calc char by char. I hesitate to change the crc calc in sumd because i don't have a method to test it, neither to debug it to see if it does the job correctly.

File: src/main/common/crc.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#define CRCCCITTANDVALUE	0x8000
```
Comment (ledvinap) [inline src/main/common/crc.h]: This is only polluting global namespace - the polynom is defined by CCITT, no need to publish it ..

And with CRCCCITTANDVALUE ... its simply topmost bit, 0x8000 is IMO easy-to-understand idiom.

File: src/main/rx/srxl.c [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    uint16_t inCrc = 0;
+
+    // Calculate on all bytes except the final two CRC bytes
+    for (i = 0; i < srxlFrameLength - 2; i++) {
+        inCrc =  crc16_CCITT(inCrc, srxlFrame[i]);
+    }
+
+    // Get the received CRC
+    crc = ((uint16_t)srxlFrame[srxlFrameLength - 2]) << 8;
+    crc = crc + ((uint16_t)srxlFrame[srxlFrameLength - 1]);
```
Comment (ledvinap) [inline src/main/rx/srxl.c]: Its easier to compute CRC of whole message (including CRC itself), resulting value will be  constant for correct message (zero when using 0 as starting value)

**Comment by saperlot** [comment]
@ledvinap makes sense, changed accordingly

**Comment by hydra** [comment]
I'm guessing that this PR and #2246 will conflict an break each other.

Can you merge #2246 locally and refactor/test as appropriate and then update this PR with the changes from #2246?

**Comment by GruffyPuffy** [comment]
@saperlot 
Please let me know if/how I can help in this matter. I think, as @hydra suggested that it might be easier to merge my additions into this PR. Basically I just changed the RJ01-part of the code. The new additions come to the fact that JR has a few versions (with slightly different protocols) of their DMSS sats. Also moved unpacking from ISR code into main code.

**Comment by saperlot** [comment]
oops my commit should be called "merge with #2246" is there a way to rename a commit, i guess no.
Anyway, @GruffyPuffy please check the merge if it works.

**Comment by hydra** [comment]
@saperlot can you rebase against master (some files moved)
@GruffyPuffy ok, can you do the testing of the new code from @saperlot and confirm correct operations.

Also, i just want to confirm that you both have tested that FC failsafe behavior works with the receivers you're using/testing.  If there is any special configuration required to make the FC failsafe work then please add that to the PR in the form user of documentation.

**Comment by GruffyPuffy** [comment]
@saperlot 
I does not work, you have removed a few critical lines... ;) Anyhow, I have a fix for it. How can I commit it to this PR? Please advice as I am not a git jedi...

@hydra I will check the failsafe. I do need to read up on how/what is supposed to happen etc. I guess we are missing something in the code?

**Comment by saperlot** [comment]
@GruffyPuffy please check if it works when you add just the defining of XBUS_FRAME_SIZE and do this change in the crc checking function: 
`for (i = 0; i < XBUS_FRAME_SIZE; i++) {`
@hydra well, will try ;-) i'm not a git expert.

**Comment by saperlot** [comment]
@GruffyPuffy i guess you have to check the failsafe with either unpowering the RX or plug out the RX signal from the controller, then it should go into failsafe configuration (if enabled). A failure would be if it still recognizes that data is coming in and failsafe won't go into failsafe in this case.

**Comment by GruffyPuffy** [comment]
@saperlot: May I suggest we use the already tested code. It has been tested with many different receivers etc.

Re failsafe: That is how it works. The receiver will output the "safe values" set while binding.

**Comment by GruffyPuffy** [comment]
@saperlot - Perfect! It works! Tested with all receivers I have.
@hydra - Failsafe works as expected, i.e. when TX is shut off the values are set by the RX/Sat to correct values.

**Comment by saperlot** [comment]
@GruffyPuffy Thanks for the quick update. Your failsafe check does not check Clmeanflight, it just tests your RX. Please see my previous comment..

**Comment by saperlot** [comment]
@GruffyPuffy oh i see you commented already my comment. Not sure where the disconnect is, but don't test the failsafe of your receiver ;-). Test the failsafe of Cleanflight by dropping the connection RX - Cleanflight board.

**Comment by GruffyPuffy** [comment]
@saperlot I looked at how the SUMD and SBUS implemented this. If I understand it correctly (cannot test) they receive a flag/bit in the data from the receiver that it is in failsafe mode. There is no such data (that I am aware of) in the xbus protocol. So, I do not think they will detect a broken RX either...or did I completely miss something? So, I do not really know what is expected from me. Shall we implement some kind of message counter and indicate failsafe if the data stream is lost. If so, I guess we need to add that to others (SRXL etc) as well. Or, maybe on a higher level (in RX)? Anyhow...getting late here. No more time today! :)

**Comment by saperlot** [comment]
@GruffyPuffy Yes you miss something. But that's no problem ;-). We should test if Cleanflight doesn't receive any data from the Receiver (after it has) that Cleanflight itself switch into failsafe. We should test if our code doesn't send just always the last same values, even if it does not receive new data.
it is already implemented with the function `xBusFrameStatus` after we disconnect the receiver from Cleanflight shall be `SERIAL_RX_FRAME_PENDING`. Cleanflight recognizes a connection loss when `SERIAL_RX_FRAME_PENDING` gets reported for a second or so. Then Cleanflight goes into failsafe. So please test it like this:
- Connect the configurator and watch the Receiver values.
- Move your sticks on the TX a bit, so that is not at default value. 
- Disconnect the RX cable (Cable between Cleanflight and your RX)
- See if the RX value on your screen changes to default values (after you disconnect the cable)

**Comment by GruffyPuffy** [comment]
@saperlot Haha...ok. I got it...but that has been in the code for a long time. It works like you described. If I for example has a high throttle value, disconnects the RX-sat, then the value (seen in the configurator) returns to 0. I.e. it does not stay with a high value. My guess is then that we are good to go regarding failsafe...unless I still miss something... :)

**Comment by saperlot** [comment]
@hydra 
- merged with master
- SRXL for failsafe tested
- awaiting failsafe test by @GruffyPuffy (i checked the code, it should work as expected)

After his test it is ready to be merged
(please also merge  https://github.com/cleanflight/cleanflight-configurator/pull/402 )

Thanks

**Comment by hydra** [comment]
@saperlot good work, thanks!

**Comment by hydra** [comment]
ok, gonna merge this now that some of the other changes are in the master branch.

**Pull request merged**